src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -106,41 +390,313 @@ public function request(string $method, string $url, array $options = []): Respo
+     * specification (RFC 9111).
+     *
+     * This function will return true if the server response can be cached,
+     * false otherwise.
+     *
+     * @param int                            $statusCode      the HTTP status code of the response
+     * @param array<string, string|string[]> $requestHeaders  the HTTP request headers
+     * @param array<string, string|string[]> $responseHeaders the HTTP response headers
+     * @param array<string, string|true>     $cacheControl    an array of parsed Cache-Control directives
+     *
+     * @return bool true if the response is cacheable, false otherwise
+     */
+    private function isServerResponseCacheable(int $statusCode, array $requestHeaders, array $responseHeaders, array $cacheControl): bool
     {
-        if ($this->client instanceof ResetInterface) {
-            $this->client->reset();
+        // no-store => skip caching
+        if (isset($cacheControl['no-store'])) {
+            return false;
         }
+
+        if (
+            $this->sharedCache
+            && !isset($cacheControl['public']) && !isset($cacheControl['s-maxage']) && !isset($cacheControl['must-revalidate'])
+            && isset($requestHeaders['authorization'])
+        ) {
+            return false;
+        }
+
+        if ($this->sharedCache && isset($cacheControl['private'])) {
+            return false;
+        }
+
+        // Conditionals require an explicit expiration
+        if (\in_array($statusCode, self::CONDITIONALLY_CACHEABLE_STATUS_CODES, true)) {
+            return $this->hasExplicitExpiration($responseHeaders, $cacheControl);
+        }
+
+        return \in_array($statusCode, self::CACHEABLE_STATUS_CODES, true);
+    }
+
+    /**
+     * Checks if the response has an explicit expiration.
+     *
+     * This function will return true if the response has an explicit expiration
+     * time specified in the headers or in the Cache-Control directives,
+     * false otherwise.
+     *
+     * @param array<string, string|string[]> $headers      the HTTP response headers
+     * @param array<string, string|true>     $cacheControl an array of parsed Cache-Control directives
+     *
+     * @return bool true if the response has an explicit expiration, false otherwise
+     */
+    private function hasExplicitExpiration(array $headers, array $cacheControl): bool
+    {
+        return isset($headers['expires'])
+            || ($this->sharedCache && isset($cacheControl['s-maxage']))
+            || isset($cacheControl['max-age']);
+    }
+
+    /**
+     * Creates a MockResponse object from cached data.
+     *
+     * This function constructs a MockResponse from the cached data, including
+     * the original request method, URL, and options, as well as the cached
+     * response headers and content. The constructed MockResponse is then
+     * returned.
+     *
+     * @param string $key        the cache key for the response
+     * @param array  $cachedData the cached data for the response
```

Comment:
this should describe the precise type for the array shape. It will benefit future contributors and maintainers reading the code, and will also make the static analysis more useful by catching potential mistakes.